src/DeterministicNoiseSimulator.cpp

```diff
@@ -106,169 +122,214 @@ std::map<std::string, double> DeterministicNoiseSimulator::DeterministicSimulate
+    dd->densityAdd.printStatistics();
+    printf("Noise cache\n");
+    dd->densityNoise.printStatistics();
     return AnalyseState(n_qubits, false);
 }
 
-qc::MatrixDD DeterministicNoiseSimulator::ApplyNoiseEffects(qc::MatrixDD density_op, const std::unique_ptr<qc::Operation>& op, unsigned char maxDepth) {
-    if (density_op.p->v < maxDepth || density_op.isTerminal()) {
-        qc::MatrixDD tmp{};
-        if (!density_op.w.approximatelyZero()) {
-            tmp.w = dd->cn.getCached(dd::CTEntry::val(density_op.w.r), dd::CTEntry::val(density_op.w.i));
-        } else {
+dEdge DeterministicNoiseSimulator::applyNoiseEffects(dEdge& originalEdge, const std::vector<dd::Qubit>& used_qubits, bool firstPathEdge) {
+    if (originalEdge.p->v < used_qubits.back() || originalEdge.isTerminal()) {
+        dEdge tmp{};
+        if (originalEdge.w.approximatelyZero()) {
             tmp.w = dd::Complex::zero;
+        } else {
+            tmp.w = originalEdge.w;
+            //            noiseLookUpResult.w = dd->cn.getCached(dd::CTEntry::val(originalEdge.w.r), dd::CTEntry::val(originalEdge.w.i));
         }
-        if (density_op.isTerminal()) {
-            return qc::MatrixDD::terminal(tmp.w);
+        if (originalEdge.isTerminal()) {
+            return dEdge::terminal(tmp.w);
         } else {
-            tmp.p = density_op.p;
+            tmp.p = originalEdge.p;
             return tmp;
         }
     }
-
-    std::vector used_qubits = op->getTargets();
-    for (auto control: op->getControls()) {
-        used_qubits.push_back(control.qubit);
-    }
-    sort(used_qubits.begin(), used_qubits.end());
-
-    std::array<qc::MatrixDD, 4> new_edges{};
-    for (int i = 0; i < 4; i++) {
-        if (density_op.p->e[i].w.approximatelyZero()) {
-            // applyNoiseEffects2 returns a pointer to the terminal node with edge weight 0
-            new_edges[i] = qc::MatrixDD::terminal(density_op.p->e[i].w);
-            continue;
-        }
-
-        // Check if the target of the current edge is in the Compute table. Note that I check for the target node of
-        // the current edge if noise needs to be applied or not
-        qc::MatrixDD tmp = noiseLookup(density_op.p->e[i], used_qubits);
-
-        if (tmp.p != nullptr) {
-            new_edges[i] = tmp;
-            continue;
+    auto originalCopy = originalEdge;
+    originalCopy.w    = dd::Complex::one;
+
+    // Check if the target of the current edge is in the "compute table".
+
+    //    auto noiseLookUpResult = dd->densityNoise.lookup(originalCopy, used_qubits);
+    //
+    //    if (noiseLookUpResult.p != nullptr) {
+    //        auto tmpComplexValue = dd->cn.getCached();
+    //        CN::mul(tmpComplexValue, noiseLookUpResult.w, originalEdge.w);
+    //        noiseLookUpResult.w = dd->cn.lookup(tmpComplexValue);
+    //        dd->cn.returnToCache(tmpComplexValue);
+    //        return noiseLookUpResult;
+    //    }
+
+    std::array<dEdge, 4> new_edges{};
```

Comment:
try to abstract away from these concrete types as much as possible.

Reply:
changed to
> std::array<dEdge, std::tuple_size_v<decltype(dd::dNode::e)>> new_edges{};